File: quality_transformer.py
#!/usr/bin/env python3
"""Optional local T5 grammar-correction adapter for the Ikmal sidecar.

The Go sidecar remains fully functional without this process. When this adapter
is running, set IKMAL_TRANSFORMER_URL to its /v1/analyze endpoint.
"""

from __future__ import annotations

import difflib
import json
import os
import re
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from typing import Any


MODEL_NAME = os.environ.get("IKMAL_TRANSFORMER_MODEL", "Unbabel/gec-t5_small")
DEFAULT_MODEL_DIR = Path.home() / ".ikmal-editor" / "models" / "gec-t5_small"
MODEL_SOURCE = os.environ.get(
    "IKMAL_TRANSFORMER_MODEL_DIR",
    str(DEFAULT_MODEL_DIR) if (DEFAULT_MODEL_DIR / "config.json").exists() else MODEL_NAME,
)
PORT = int(os.environ.get("IKMAL_TRANSFORMER_PORT", "8099"))
MAX_BODY = 1 << 20
MAX_CHUNK_WORDS = int(os.environ.get("IKMAL_TRANSFORMER_MAX_CHUNK_WORDS", "80"))
MAX_INPUT_TOKENS = int(os.environ.get("IKMAL_TRANSFORMER_MAX_INPUT_TOKENS", "128"))
TOKEN_RE = re.compile(r"\w+|[^\w\s]", re.UNICODE)
WORD_RE = re.compile(r"\w+", re.UNICODE)

_tokenizer: Any = None
_model: Any = None
_torch: Any = None
_load_error: str | None = None


def load_model() -> tuple[Any, Any, Any]:
    global _tokenizer, _model, _torch, _load_error
    if _tokenizer is not None:
        return _tokenizer, _model, _torch
    if _load_error is not None:
        raise RuntimeError(_load_error)
    try:
        import torch
        from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

        _torch = torch
        _tokenizer = AutoTokenizer.from_pretrained(MODEL_SOURCE)
        _model = AutoModelForSeq2SeqLM.from_pretrained(MODEL_SOURCE)
        _model.eval()
        return _tokenizer, _model, _torch
    except Exception as exc:  # dependency and model-download failures are reported by /health
        _load_error = f"{type(exc).__name__}: {exc}"
        raise RuntimeError(_load_error) from exc


def correct(text: str) -> str:
    tokenizer, model, torch = load_model()
    prompt = "gec: " + text
    encoded = tokenizer(prompt, return_tensors="pt", truncation=True, max_length=MAX_INPUT_TOKENS)
    with torch.no_grad():
        output = model.generate(
            **encoded,
            max_length=MAX_INPUT_TOKENS,
            num_beams=5,
            early_stopping=True,
        )
    return tokenizer.decode(output[0], skip_special_tokens=True, clean_up_tokenization_spaces=True)


def tokenize(text: str) -> list[tuple[str, int, int]]:
    return [(match.group(0), match.start(), match.end()) for match in TOKEN_RE.finditer(text)]


def detokenize(parts: list[str]) -> str:
    value = " ".join(parts)
    value = re.sub(r"\s+([,.!?;:%])", r"\1", value)
    value = re.sub(r"\s+(['’])", r"\1", value)
    value = re.sub(r"(['’])\s+", r"\1", value)
    return value


def utf16_offset(text: str, codepoint_offset: int) -> int:
    return len(text[:codepoint_offset].encode("utf-16-le")) // 2


def suggestions_for(text: str, corrected: str) -> list[dict[str, Any]]:
    if text == corrected:
        return []

    source = tokenize(text)
    target = [part[0] for part in tokenize(corrected)]
    source_values = [part[0] for part in source]
    matcher = difflib.SequenceMatcher(a=source_values, b=target, autojunk=False)
    suggestions = []

    for tag, i1, i2, j1, j2 in matcher.get_opcodes():
        if tag == "equal":
            continue
        if i1 < len(source):
            start = source[i1][1]
            end = source[i2 - 1][2] if i2 > i1 else start
        else:
            start = end = len(text)
        replacement = detokenize(target[j1:j2])
        suggestions.append(
            {
                "start": utf16_offset(text, start),
                "end": utf16_offset(text, end),
                "replacement": replacement,
                "category": "transformer-grammar",
                "message": "The local grammar model suggests this correction.",
                "confidence": 0.68,
                "source": "transformer",
            }
        )
    return suggestions


def sentence_ranges(text: str) -> list[tuple[int, int]]:
    ranges = []
    start = 0
    for token, _, end in tokenize(text):
        if token in ".!?":
            ranges.append((start, end))
            start = end
    if start < len(text):
        ranges.append((start, len(text)))
    return ranges


def word_count(text: str) -> int:
    return len(WORD_RE.findall(text))


def split_long_range(text: str, start: int, end: int) -> list[tuple[int, int]]:
    words = list(WORD_RE.finditer(text, start, end))
    if not words:
        return [(start, end)]
    ranges = []
    chunk_start = start
    for index, word in enumerate(words, start=1):
        if index % MAX_CHUNK_WORDS == 0:
            ranges.append((chunk_start, word.end()))
            chunk_start = words[index].start() if index < len(words) else word.end()
    if chunk_start < end:
        ranges.append((chunk_start, end))
    return ranges


def correction_chunks(text: str) -> list[tuple[int, int]]:
    """Group complete sentences without exceeding the local model budget."""
    ranges = []
    pending_start = None
    pending_end = None
    pending_words = 0

    def flush_pending() -> None:
        nonlocal pending_start, pending_end, pending_words
        if pending_start is not None and pending_end is not None:
            ranges.append((pending_start, pending_end))
        pending_start = None
        pending_end = None
        pending_words = 0

    for start, end in sentence_ranges(text):
        count = word_count(text[start:end])
        if count > MAX_CHUNK_WORDS:
            flush_pending()
            ranges.extend(split_long_range(text, start, end))
            continue
        if pending_start is not None and pending_words + count > MAX_CHUNK_WORDS:
            flush_pending()
        if pending_start is None:
            pending_start = start
        pending_end = end
        pending_words += count
    flush_pending()
    return ranges


def add_utf16_offset(text: str, base: int, value: int) -> int:
    return utf16_offset(text, base) + value


def analyze_chunk(text: str, start: int, end: int) -> list[dict[str, Any]]:
    chunk = text[start:end]
    corrected = correct(chunk)
    suggestions = suggestions_for(chunk, corrected)
    for suggestion in suggestions:
        suggestion["start"] = add_utf16_offset(text, start, suggestion["start"])
        suggestion["end"] = add_utf16_offset(text, start, suggestion["end"])
    return suggestions


def analyze(payload: dict[str, Any]) -> dict[str, Any]:
    text = payload.get("text", "")
    if not isinstance(text, str) or not text.strip():
        raise ValueError("text is required")
    suggestions = []
    for start, end in correction_chunks(text):
        suggestions.extend(analyze_chunk(text, start, end))
    return {
        "backend": "transformer",
        "suggestions": suggestions,
        "antecedents": [],
    }


class Handler(BaseHTTPRequestHandler):
    def send_json(self, status: int, value: dict[str, Any]) -> None:
        body = json.dumps(value).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self) -> None:  # noqa: N802
        if self.path != "/health":
            self.send_json(404, {"error": "not found"})
            return
        self.send_json(
            200,
            {
                "status": "ok" if _load_error is None else "degraded",
                "backend": "transformer",
                "model": MODEL_SOURCE,
                "loaded": _tokenizer is not None,
                "error": _load_error,
            },
        )

    def do_POST(self) -> None:  # noqa: N802
        if self.path != "/v1/analyze":
            self.send_json(404, {"error": "not found"})
            return
        try:
            length = int(self.headers.get("Content-Length", "0"))
            if length > MAX_BODY:
                raise ValueError("request too large")
            payload = json.loads(self.rfile.read(length))
            self.send_json(200, analyze(payload))
        except Exception as exc:
            self.send_json(503, {"error": str(exc)})

    def log_message(self, format: str, *args: Any) -> None:
        return


if __name__ == "__main__":
    server = ThreadingHTTPServer(("127.0.0.1", PORT), Handler)
    print(f"Ikmal transformer adapter listening on http://127.0.0.1:{PORT}")
    server.serve_forever()

File: quality_transformer.mjs
#!/usr/bin/env node

import http from 'node:http';
import os from 'node:os';
import path from 'node:path';
import process from 'node:process';

const PORT = Number(process.env.IKMAL_TRANSFORMER_PORT || 8099);
const MAX_BODY = 1 << 20;
const MAX_CHUNK_WORDS = Number(process.env.IKMAL_TRANSFORMER_MAX_CHUNK_WORDS || 80);
const MODEL_ID = process.env.IKMAL_TRANSFORMER_MODEL || 'Xenova/t5-base-grammar-correction';
const MODEL_CACHE_DIR =
  process.env.IKMAL_TRANSFORMER_CACHE_DIR || path.join(os.homedir(), '.ikmal-editor', 'models');
// This model publishes q4 ONNX files, but not q4f16 files. q4 is also the
// quantization mode used by ikmal's WASM Transformers.js tier.
const DTYPE = process.env.IKMAL_TRANSFORMER_DTYPE || 'q4';

let transformers = null;
let generatorPromise = null;
let loadError = null;

async function loadGenerator() {
  if (generatorPromise) return generatorPromise;
  generatorPromise = (async () => {
    transformers = await import('@huggingface/transformers');
    transformers.env.cacheDir = MODEL_CACHE_DIR;
    transformers.env.localModelPath = MODEL_CACHE_DIR;
    return transformers.pipeline('text2text-generation', MODEL_ID, {
      dtype: DTYPE,
      progress_callback: (event) => {
        if (event?.status === 'progress') {
          process.stdout.write(`\r${event.file || 'model'} ${Math.round(event.progress || 0)}%`);
        }
      },
    });
  })().catch((error) => {
    loadError = `${error?.name || 'Error'}: ${error?.message || error}`;
    generatorPromise = null;
    throw error;
  });
  return generatorPromise;
}

function tokenize(text) {
  const pattern = /\w+|[^\w\s]/gu;
  return [...text.matchAll(pattern)].map((match) => ({
    value: match[0],
    start: match.index,
    end: match.index + match[0].length,
  }));
}

function detokenize(parts) {
  return parts
    .join(' ')
    .replace(/\s+([,.!?;:%])/g, '$1')
    .replace(/\s+(['’])/g, '$1')
    .replace(/(['’])\s+/g, '$1');
}

function wordCount(text) {
  return [...text.matchAll(/[\p{L}\p{N}_]+/gu)].length;
}

function sentenceRanges(text) {
  const ranges = [];
  let start = 0;
  for (const token of tokenize(text)) {
    if ('.!?'.includes(token.value)) {
      ranges.push([start, token.end]);
      start = token.end;
    }
  }
  if (start < text.length) ranges.push([start, text.length]);
  return ranges;
}

function splitLongRange(text, start, end) {
  const words = [...text.slice(start, end).matchAll(/[\p{L}\p{N}_]+/gu)].map((match) => ({
    start: start + match.index,
    end: start + match.index + match[0].length,
  }));
  if (!words.length) return [[start, end]];
  const ranges = [];
  let chunkStart = start;
  for (let index = MAX_CHUNK_WORDS; index <= words.length; index += MAX_CHUNK_WORDS) {
    ranges.push([chunkStart, words[index - 1].end]);
    chunkStart = index < words.length ? words[index].start : words[index - 1].end;
  }
  if (chunkStart < end) ranges.push([chunkStart, end]);
  return ranges;
}

function correctionChunks(text) {
  const ranges = [];
  let pending = null;
  let pendingWords = 0;
  const flush = () => {
    if (pending) ranges.push(pending);
    pending = null;
    pendingWords = 0;
  };

  for (const [start, end] of sentenceRanges(text)) {
    const count = wordCount(text.slice(start, end));
    if (count > MAX_CHUNK_WORDS) {
      flush();
      ranges.push(...splitLongRange(text, start, end));
      continue;
    }
    if (pending && pendingWords + count > MAX_CHUNK_WORDS) flush();
    if (!pending) pending = [start, end];
    else pending[1] = end;
    pendingWords += count;
  }
  flush();
  return ranges;
}

function diffSuggestions(text, corrected, baseOffset) {
  const source = tokenize(text);
  const target = tokenize(corrected).map((token) => token.value);
  const sourceValues = source.map((token) => token.value);
  const rows = sourceValues.length + 1;
  const cols = target.length + 1;
  const lcs = Array.from({ length: rows }, () => new Uint16Array(cols));
  for (let i = sourceValues.length - 1; i >= 0; i -= 1) {
    for (let j = target.length - 1; j >= 0; j -= 1) {
      lcs[i][j] = sourceValues[i] === target[j] ? lcs[i + 1][j + 1] + 1 : Math.max(lcs[i + 1][j], lcs[i][j + 1]);
    }
  }

  const suggestions = [];
  let i = 0;
  let j = 0;
  while (i < sourceValues.length || j < target.length) {
    if (i < sourceValues.length && j < target.length && sourceValues[i] === target[j]) {
      i += 1;
      j += 1;
      continue;
    }
    const sourceStart = i;
    const targetStart = j;
    while (i < sourceValues.length || j < target.length) {
      if (i < sourceValues.length && j < target.length && sourceValues[i] === target[j]) break;
      if (j < target.length && (i === sourceValues.length || lcs[i][j + 1] >= lcs[i + 1][j])) j += 1;
      else i += 1;
    }
    const start = sourceStart < source.length ? source[sourceStart].start : text.length;
    const end = sourceStart < source.length ? source[Math.max(sourceStart, i) - 1].end : start;
    const replacement = detokenize(target.slice(targetStart, j));
    // Keep this adapter focused on grammar corrections. The small T5 model
    // occasionally rewrites a sentence boundary (for example, joining two
    // sentences with "and the"); that is a style rewrite, not a safe grammar
    // suggestion for this endpoint.
    if (/[.!?]/.test(text.slice(start, end)) && !/[.!?]/.test(replacement)) continue;
    suggestions.push({
      start: baseOffset + start,
      end: baseOffset + end,
      replacement,
      category: 'transformer-grammar',
      message: 'The local grammar model suggests this correction.',
      confidence: 0.68,
      source: 'transformer-js-onnx',
    });
  }
  return suggestions;
}

async function analyze(payload) {
  if (!payload || typeof payload.text !== 'string' || !payload.text.trim()) {
    throw new Error('text is required');
  }
  const generator = await loadGenerator();
  const suggestions = [];
  for (const [start, end] of correctionChunks(payload.text)) {
    const output = await generator(payload.text.slice(start, end), {
      max_new_tokens: 128,
      num_beams: 4,
      do_sample: false,
    });
    const corrected = output?.[0]?.generated_text || payload.text.slice(start, end);
    suggestions.push(...diffSuggestions(payload.text.slice(start, end), corrected, start));
  }
  return { backend: 'transformer-js-onnx', suggestions, antecedents: [] };
}

function sendJSON(response, status, value) {
  const body = JSON.stringify(value);
  response.writeHead(status, { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) });
  response.end(body);
}

const server = http.createServer(async (request, response) => {
  if (request.method === 'GET' && request.url === '/health') {
    sendJSON(response, 200, {
      status: loadError ? 'degraded' : 'ok',
      backend: 'transformer-js-onnx',
      model: MODEL_ID,
      dtype: DTYPE,
      cacheDir: MODEL_CACHE_DIR,
      loaded: Boolean(generatorPromise && !loadError),
      error: loadError,
    });
    return;
  }
  if (request.method !== 'POST' || request.url !== '/v1/analyze') {
    sendJSON(response, 404, { error: 'not found' });
    return;
  }
  let body = '';
  request.on('data', (chunk) => {
    body += chunk;
    if (Buffer.byteLength(body) > MAX_BODY) request.destroy();
  });
  request.on('end', async () => {
    try {
      sendJSON(response, 200, await analyze(JSON.parse(body)));
    } catch (error) {
      sendJSON(response, 503, { error: error?.message || String(error) });
    }
  });
});

if (process.argv.includes('--preload')) {
  loadGenerator()
    .then(() => {
      console.log(`Quality model ready: ${MODEL_ID}`);
      process.exit(0);
    })
    .catch(() => process.exit(1));
} else {
  server.listen(PORT, '127.0.0.1', () => {
    console.log(`ikmal transformer adapter listening on http://127.0.0.1:${PORT}`);
  });
}
